dot/state/pruner.go

```diff
@@ -0,0 +1,274 @@
+	p.Lock()
+	defer p.Unlock()
+
+	// remove re-inserted keys
+	for _, k := range jr.insertedKeys {
+		if num, ok := p.deathIndex[k]; ok {
+			delete(p.deathList[num-p.pendingNumber].deletedKeys, k)
+			delete(p.deathIndex, k)
+		}
+	}
+
+	// add deleted keys from journal to death index
+	for _, k := range jr.deletedKeys {
+		p.deathIndex[k] = blockNum
+	}
+
+	deletedKeys := make(map[*common.Hash]int64)
+	for _, data := range jr.deletedKeys {
+		deletedKeys[data] = blockNum
+	}
+
+	dr := &deathRow{
+		blockHash:   jr.blockHash,
+		deletedKeys: deletedKeys,
+	}
+
+	// add deathRow to deathList
+	p.deathList = append(p.deathList, dr)
+	return nil
+}
+
+func (p *FullNodePruner) start() {
+	logger.Info("pruning started")
+
+	for {
+		p.Lock()
+		if int64(len(p.deathList)) <= p.retainBlocks {
+			p.Unlock()
+			time.Sleep(2 * time.Second)
+			continue
+		}
+
+		logger.Info("pruner", "pruning block ", p.pendingNumber)
+
+		// pop first element from death list
+		row := p.deathList[0]
+		err := p.deleteKeys(row.deletedKeys)
+		if err != nil {
+			logger.Error("pruner", "failed to delete keys for block", p.pendingNumber)
+			continue
+		}
+
+		for k := range row.deletedKeys {
+			delete(p.deathIndex, k)
+		}
+
+		err = p.storeLastPrunedIndex(p.pendingNumber)
+		if err != nil {
+			logger.Error("pruner", "failed to store last pruned index")
+		}
+
+		p.deathList = p.deathList[1:]
+		p.pendingNumber++
+		p.Unlock()
+	}
+}
+
+func (p *FullNodePruner) storeJournal(blockNum int64, jr *journalRecord) error {
+	encRecord, err := scale.Encode(jr)
+	if err != nil {
+		return fmt.Errorf("failed to encode journal record %d: %w", blockNum, err)
+	}
+
+	encNum, err := scale.Encode(blockNum)
+	if err != nil {
+		return err
+	}
+
+	err = p.journalDB.Put(encNum, encRecord)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *FullNodePruner) getJournalRecord(num int64) (*journalRecord, error) {
+	encNum, err := scale.Encode(num)
+	if err != nil {
+		return nil, err
+	}
+
+	val, err := p.journalDB.Get(encNum)
+	if err != nil {
+		return nil, err
+	}
+
+	decJR, err := scale.Decode(val, new(journalRecord))
+	if err != nil {
+		return nil, err
+	}
+
+	return decJR.(*journalRecord), nil
+}
+
+func (p *FullNodePruner) storeLastPrunedIndex(blockNum int64) error {
+	encNum, err := scale.Encode(blockNum)
+	if err != nil {
+		return err
+	}
+
+	err = p.journalDB.Put([]byte(lastPrunedKey), encNum)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *FullNodePruner) getLastPrunedIndex() (int64, error) {
+	val, err := p.journalDB.Get([]byte(lastPrunedKey))
+	if err == chaindb.ErrKeyNotFound {
+		return 0, nil
+	}
+
+	if err != nil {
+		return 0, err
+	}
+
+	blockNum, err := scale.Decode(val, int64(0))
+	if err != nil {
+		return 0, err
+	}
+
+	return blockNum.(int64), err
+}
+
+func (p *FullNodePruner) deleteKeys(nodesHash map[*common.Hash]int64) error {
+	for k := range nodesHash {
+		err := p.storageDB.Del(k.ToBytes())
+		if err != nil {
+			return err
+		}
```

Comment:
continue instead of return

Reply:
I was thinking that the pruner should not move to the next block in case of any error. This error would only happen in case of DB corruption or wrong code logic.